### PR TITLE
[3] Add ReductionOverNeighbor - Seperate

### PR DESCRIPTION
**Pull request by lukasm91**
See https://github.com/MeteoSwiss-APN/dawn/pull/269

### CODE REVIEW

File: src/dawn/AST/ASTExpr.cpp [hunk 1]
```diff
@@ -430,5 +431,33 @@ bool LiteralAccessExpr::equals(const Expr* other) const {
   return otherPtr && Expr::equals(other) && value_ == otherPtr->value_ &&
          builtinType_ == otherPtr->builtinType_;
 }
+
+ReductionOverNeighborExpr::ReductionOverNeighborExpr(std::string const& op,
+                                                     std::shared_ptr<Expr> rhs,
+                                                     std::shared_ptr<Expr> init, SourceLocation loc)
+    : Expr(EK_ReductionOverNeighborExpr, loc), op_(op), rhs_(std::move(rhs)),
```
Comment (havogt) [inline src/dawn/AST/ASTExpr.cpp]: I think the moves disable copy-elision.

File: src/dawn/AST/ASTStringifier.cpp [hunk 2]
```diff
@@ -143,6 +143,11 @@ class StringVisitor : public ASTVisitor {
       stmt->getElseStmt()->accept(*this);
     }
   }
+  void visit(const std::shared_ptr<ReductionOverNeighborExpr>& expr) override {
+    if(scopeDepth_ == 0)
+      ss_ << std::string(curIndent_, ' ');
```
Comment (havogt) [inline src/dawn/AST/ASTStringifier.cpp]: How does it related to scopeDepth==0?

Comment (lukasm91) [reply src/dawn/AST/ASTStringifier.cpp]: I have no idea, what and how this is relevant for Stmts, but obviously it is not for Expr. THese lines were outdated anyway. Fixed.

File: src/dawn/AST/ASTExpr.cpp (hunk 1 above)
Comment (lukasm91) [reply src/dawn/AST/ASTExpr.cpp]: I am not sure (I know it's a problem in the return, bu t not sure about constructores). As this was copied from another place, I replaced it with `const&` now

Comment (havogt) [reply src/dawn/AST/ASTExpr.cpp]: I verified in a small snippet that copy elision is indeed disabled in this case. So the move should be avoided.

Comment (havogt) [reply src/dawn/AST/ASTExpr.cpp]: In a different branch? Or didn't push?

Comment (havogt) [reply src/dawn/AST/ASTExpr.cpp]: Ah clicked on view changes and was sent to the old state.

**Review by havogt — APPROVED** [review]

**Pull request closed without merging**